ACCS-366: My Account: Stored payment methods page (#816)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/user-account/containers/payment-methods.mdx

---
title: PaymentMethods container
description: Learn about the PaymentMethods container for saved payment methods in My Account.
---

import Diagram from '@components/Diagram.astro';
import OptionsTable from '@components/OptionsTable.astro';
import Aside from '@components/Aside.astro';

> The `PaymentMethods` container shows saved payment methods for the signed-in customer (for example, cards saved at checkout). **Vaulting** means Adobe Commerce stores a **token** for the card (not the full card number) so the shopper can reuse that method later. You can show a full section or a **compact** row layout next to other account widgets. The list uses the same card styling as checkout flows through the `PaymentCard` component.

Reviewer: ```suggestion
The `PaymentMethods` container displays saved payment methods for the signed-in customer (for example, cards saved during checkout). Vaulting means Adobe Commerce stores a token—not the full card number—so the shopper can reuse the method later. You can render it as a full section or a compact layout alongside other account widgets. The list uses the same card styling as checkout via the `PaymentCard` component.
```

## Data loading

The container uses the `usePaymentMethods` hook in the `@dropins/storefront-account` package, which chooses a data source as follows:

> 1. **Event bus (eager)** — The hook subscribes to `account/customerPaymentTokens` on `events.on` with the listener option `eager` set to `true` so any existing payload is applied immediately. If an event payload is already available, the **display list** is taken from the **event bus** and **no** GraphQL load runs on mount. This path is **injection only**; it does not define how removal works.

Reviewer: ```suggestion
1. Event bus (eager) — The hook subscribes to the `account/customerPaymentTokens` via `events.on` with `eager: true`, so any existing payload is applied immediately. If a payload is already present, the display list is sourced from the event bus, and no GraphQL request runs on mount. This path is for injection only and does not define removal behavior.
```

> 2. **GraphQL** — If nothing is consumed from the bus, the hook calls **`getCustomerPaymentTokens`**.

Reviewer: ```suggestion
2. **GraphQL** — If no data is received from the event bus, the hook calls `getCustomerPaymentTokens`.
```

> **Remove** always uses the GraphQL **`deletePaymentToken`** mutation with the token's **public_hash** (never the event bus), then updates the list on success—whether the rows were first shown from the bus or from **`getCustomerPaymentTokens`**.

Reviewer: ```suggestion
The **Remove** button always calls the GraphQL `deletePaymentToken` mutation with the token’s `public_hash` (never the event bus), then updates the list on success—regardless of whether the rows were initially sourced from the bus or from `getCustomerPaymentTokens`.
```

> In the default **PaymentMethods** container, the shopper must **confirm in the removal modal** before the hook runs that delete (or the local-only update when the list came from the event bus).

Reviewer: ```suggestion
In the default `PaymentMethods` container, the shopper must confirm in the removal modal before the hook executes the delete (or performs a local-only update if the list was sourced from the event bus).
```

> Changing `filterPaymentMethodCodes` after the initial load triggers a refetch **only** when the current source is GraphQL (not when data came from the event bus).

Reviewer: ```suggestion
Changing `filterPaymentMethodCodes` after the initial load triggers a refetch only when the data source is GraphQL (not when the data comes from the event bus).
```

For the **`account/customerPaymentTokens`** payload type in TypeScript, see [TypeScript declaration](/dropins/user-account/events/#typescript-declaration) on the Events page.

<Aside type="tip" title="Injecting tokens from another drop-in">

> Emit the **account/customerPaymentTokens** event with an array of display rows (the same shape as the **StoredPaymentMethodDisplay** model in the drop-in source) when you already have token data on the client—for example, after checkout or Payment Services initialization. Emit a **null** payload for an explicit empty list if your integration follows the same convention as the cart **cart/data** event. That event only **feeds the list**; **Remove** in the UI still runs **`deletePaymentToken`** via GraphQL. Re-emit an updated array after removal if other listeners need to match the new vault state.

Reviewer: ```suggestion
Emit the `account/customerPaymentTokens` event with an array of display rows (matching the `StoredPaymentMethodDisplay` model) when token data is already available on the client—for example, after checkout or Payment Services initialization. Emit a `null` payload to represent an explicit empty list if your integration follows the same convention as the cart `cart/data` event. This event only populates the list. The **Remove** button in the UI still calls `deletePaymentToken` via GraphQL. Re-emit an updated array after removal if other listeners need to stay in sync with the latest vault state.
```

</Aside>

## Layout

### Full layout

> Use the default layout on a dedicated **Payments** or account settings page: section header (unless disabled), list of cards with remove actions, and full empty state when there are no saved methods.

Reviewer: ```suggestion
Use the default layout on a dedicated Payments or account settings page. Include a section header (unless disabled), a list of cards with remove actions, and a full empty state when no saved methods exist.
```

<Diagram caption="Full layout with multiple saved cards">
  ![Full layout with multiple saved cards](@images/dropins/user-account/payment-methods-full.png)
</Diagram>

### Minified layout

> Set **`minifiedView`** when the container sits beside other dashboard widgets. The empty state and each card use a single-row layout (brand, masked number, optional expired tag, Remove).

Reviewer: ```suggestion
Set `minifiedView` when placing the container alongside other dashboard widgets. In this mode, both the empty state and each card use a single-row layout (brand, masked number, optional expired tag, and the **Remove** button).
```

<Diagram caption="Minified layout with multiple saved cards">
  ![Minified layout with multiple saved cards](@images/dropins/user-account/payment-methods-minified.png)
</Diagram>

## Remove confirmation

> This section describes **what the shopper sees** when they remove a **stored payment method** from the list—for example, a card they saved (vaulted) at checkout—and **what runs on the server** afterward, so you can test the flow or explain it to authors.

Reviewer: ```suggestion
This section describes what the shopper sees when removing a stored payment method (for example, a card saved at checkout) and what happens on the server afterward, so you can test the flow or explain it to authors.
```

1. The shopper selects **Remove** on a row. 

> 2. A **remove confirmation** dialog opens. The dialog can show a small **PaymentCard** preview with the card brand, last digits, and expired state when that data is available.

Reviewer: ```suggestion
2. A remove confirmation dialog opens. When displayed, the dialog may include a small `PaymentCard` preview showing the card brand, last digits, and expired status.
```

The shopper then chooses one option:

- **Cancel** closes the dialog. Nothing is deleted on the server.

> - **Remove** applies the deletion. When the list was loaded with GraphQL, the UI calls **`deletePaymentToken`**. When the list came only from the event bus, only the in-memory row is removed (see [Data loading](#data-loading)).

Reviewer: ```suggestion
- The **Remove** button applies the deletion. If the list was loaded via GraphQL, the UI calls `deletePaymentToken`. If the list came only from the event bus, it removes the item in memory (see [Data loading](#data-loading)).
```

When **Remove** succeeds, the row disappears from the list and the dialog closes.
### Full layout

<Diagram caption="Remove payment confirmation modal (full Payments page)">
  ![Remove payment confirmation modal on a full Payments page](@images/dropins/user-account/remove-payment-confirmation-modal.png)
</Diagram>

### Minified layout

<Diagram caption="Remove payment confirmation modal (minified dashboard)">
  ![Remove payment confirmation modal in minified layout](@images/dropins/user-account/remove-payment-confirmation-modal-minified.png)
</Diagram>

While a removal request is running, the dialog shows **loading** on the confirm control. **Remove** is disabled on every row until the request finishes. 

You can change the dialog labels and body text with dictionary keys under **`Account.Payments.removePaymentModal`** (see [Stored payment methods (Payments)](/dropins/user-account/dictionary/#stored-payment-methods-payments)).

## Configurations

> The `PaymentMethods` container accepts standard HTML div attributes plus the following options:

Reviewer: ```suggestion
The `PaymentMethods` container accepts standard HTML `div` attributes plus the following options:
```
